reference/JSDoc-Type-References.md

```diff
@@ -0,0 +1,242 @@
+In Javascript files, the compiler understands Typescript types in
+JSDoc comments like `@type` and `@param`. However, it has additional
+rules to make type resolution work better for Javascript code that
+wasn't written with machine checking in mind. These rules apply only
+in JSDoc in Javascript files, and some can be disabled by setting
+`"noImplicitAny": true`, which the compiler takes as a signal that the
+code is being written with Typescript in mind.
+
+This document first describes the way Javascript type resolution is
+different, because those differences are bigger and more surprising.
+Then it covers differences in name resolution.
+
+## Simple Rewrites ##
+
+The simplest Javascript-only rule is to rewrite the primitive object
+types to real primitives.
+
+Type        | Resolved Type
+------------|--------------
+`Number`    | `number`
+`String`    | `string`
+`Boolean`   | `boolean`
+`Object`    | `any`
+
+The rewrite `Object -> any` is disabled when `"noImplicitAny": true`,
+which the compiler takes as a signal that the code is being written
+with Typescript in mind. Old versions of Typescript also rewrote
+`object -> any`, but stopped in Typescript 3.7.
+
+There are also rewrites of other JSDoc types that are equivalent to
+built-in Typescript types:
+
+Type                  | Resolved Type
+----------------------|--------------
+`Null`                | `null`
+`Undefined`           | `undefined`
+`Void`                | `void`
+`function`            | `Function`
+`Object<string,Type>` | `{ [x: string]: Type }`
+`Object<number,Type>` | `{ [x: number]: Type }`
+
+Finally, `Array` and `Promise` both default their type parameters to
+`any` when no type parameters are provided:
+
+Type        | Resolved Type
+------------|--------------
+`Array`     | `any[]`
+`Promise`   | `Promise<any>`
+
+This is standard for all type references in JSDoc, so I don't think
+it's needed anymore. It only applies when `"noImplicitAny": false`.
+
+### Where To Find The Code ###
+
+`getIntendedJSDocTypeReference` in src/compiler/checker.ts.
+
+## Values as Types ##
+
+The compiler maintains three namespaces: one each for values, types and
+namespaces. Put another way, each name can have a value meaning, a
+type meaning and a namespace meaning. However, there are fewer ways
+to declare and reference types in Javascript, and the distinction
+between type and value can be subtle, so for Javascript, the compiler
+tries to resolve type references as values if it can't find a type.
+
+Here's an example:
+
+```ts
+const FOO = "foo"
+const BAR = "bar"
+
+/** @param {FOO | BAR} type */
+function f(type) {
+}
+```
+
+In the compiler, this is resolved in two stages. Both stages use a
+JSDoc fallback. First the compiler resolves the name; then it finds
+the type for the name. Normally type resolution only looks in the type
+namespace, but when it finds nothing there, a fallback looks in the
+value namespace.
```

Comment:
I'll try giving code samples and see how that looks. I've avoided it in the past but it seems like it might help here.

Reply:
To clarify, I was talking pointing at the place in the code where the fallback is implemented since it's not easy to see.  This might not be the place to point there though...

Reply:
Done to here, I'll continue tomorrow.